sematic/examples/lightning_resnet/checkpointing.py

```diff
@@ -0,0 +1,77 @@
+# Standard Library
+import io
+import logging
+from dataclasses import dataclass
+
+# Third-party
+import boto3
+import cloudpickle
+from pytorch_lightning.plugins.io.checkpoint_plugin import CheckpointIO
+
+# Sematic
+from sematic.types.types.aws.s3 import S3Location
+from sematic.utils.retry import retry
+
+logger = logging.getLogger(__name__)
+
+
+class SematicCheckpointIO(CheckpointIO):
+    """Example checkpointer for usage with PyTorch Lightning & S3 storage.
+
+    You can use the built-in PyTorch Lightning checkpointer with S3 (AWS),
+    GCS (Google Cloud), or ADL (Azure) provided that you add extra dependencies.
+    See Lightning Documentation:
+    https://pytorch-lightning.readthedocs.io/en/stable/common/checkpointing_advanced.html
+    """
+
+    def __init__(self, *args, s3_location=None, **kwargs) -> None:
+        if s3_location is None:
+            raise ValueError("Missing required argument 's3_location'")
+        self._s3_location = s3_location
+        super().__init__(*args, **kwargs)
+
+    def save_checkpoint(self, checkpoint, path, storage_options=None):
+        as_bytes = cloudpickle.dumps(checkpoint)
+        s3_client = boto3.client("s3")
+        key = (self._s3_location / path).location
+        with io.BytesIO(as_bytes) as file_obj:
+            s3_client.upload_fileobj(file_obj, self._s3_location.bucket.name, key)
+
+    @retry(tries=3, delay=5)
+    def load_checkpoint(self, path, storage_options=None):
+        file_obj = io.BytesIO()
+        s3_client = boto3.client("s3")
+        key = (self._s3_location / path).location
+        logger.info("Loading checkpoint from %s", (self._s3_location / path).to_uri())
+        s3_client.download_fileobj(self._s3_location.bucket.name, key, file_obj)
+        return cloudpickle.loads(file_obj.getvalue())
+
+    def remove_checkpoint(self, path):
+        logger.warning(
+            f"Cannot remove checkpoints once "
+            f"they have been written. Ignoring removal of: {path}"
+        )
+
+    def from_path(self, path):
+        return Checkpoint(
+            location=self._s3_location / path,
+            prefix=self._s3_location,
+        )
+
+
+@dataclass
+class Checkpoint:
+    """Sematic representation of the Checkpoint.
+
+    Attributes
+    ----------
+    location: The full location of the checkpoint
+    prefix: The prefix the checkpointer was using when the checkpoint was created
+    """
+
+    location: S3Location
+    prefix: S3Location
+
+    @property
+    def path(self):
+        return self.location.location.replace(f"{self.prefix.location}/", "")
```

Comment:
Can't you just always use the entire path and avoid re-constituting it later with `self._s3_location / path`?

Reply:
These are the options for what `Checkpoint` stores, as I see it:

1. prefix + relative path: doesn't auto-render a link to the final checkpoint location in the UI
2. full path: doesn't distinguish what's part of the prefix and what's part of the relative path. Why is that important? Because the evaluator needs to be able to know what prefix was used in order to re-instantiate the `SematicCheckpointIO` to pass to lightning's evaluator, and the prefix can only be determined by info available at runtime (specifically, it uses the run id of the train run)
3. prefix + full path: requires reconstructing the relative path from the two pieces of info stored

3 seemed like the best option to me.

Reply:
Thank you for clarifying! Please add this information as a code comment.